### PR TITLE
[WIP] Feature/presence

### PR DESCRIPTION
**Pull request by AlexBHarley**

### CODE REVIEW

File: src/message/connection-endpoint.js [hunk 1]
```diff
@@ -376,6 +380,12 @@ ConnectionEndpoint.prototype._registerAuthenticatedSocket  = function( socketWra
 		socketWrapper.sendMessage( C.TOPIC.AUTH, C.ACTIONS.ACK, [ messageBuilder.typed( userData.clientData ) ] );
 	}
 
+	if( userData.username != 'OPEN' ) {
+		var msg = messageBuilder.getMsg( C.TOPIC.PRESENCE, C.ACTIONS.PRESENCE_ADD, [ messageBuilder.typed( userData.username ) ] );
+		debugger;
+		setTimeout( this.onMessage( socketWrapper, msg ), 0);
```
Comment (yasserf) [inline src/message/connection-endpoint.js]: in node you can use

`process.nextTick( function );`

which is more efficient

File: src/presence/presence-handler.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+			this._loggedInClients.remove( messageParser.convertTyped( message.data[ 0 ] ) );
+		}
+		else if( message.action === C.ACTIONS.QUERY ) {			
+			var clients = this._loggedInClients.getAll();
+			socketWrapper.sendMessage( C.TOPIC.PRESENCE, C.ACTIONS.QUERY, clients );
+		}
+	}
+
+	_onClientAdded( username ) {
+		var sockWrappers = this._connectionEndpoint.getAuthenticatedSockets();
```
Comment (yasserf) [inline src/presence/presence-handler.js]: So we should only send to clients who are interested..

That means:

-> client does a subscribe to presense
-> presence handlers get subscribe, adds to subscription registry
-> when a client is added / removed, it sends it to those in the subscription registry

The benefit of this is the fact that we limit subscriptions only to those who are interested

File: src/message/connection-endpoint.js (hunk 1 above)
Comment (yasserf) [inline src/message/connection-endpoint.js]: We might also need to actually call the method directly on the presense, i been thinking about this approach and it can easily allow users to trigger the same event using that topic and message which would break the state of the system. See deepstream.io comment

File: src/deepstream.io.js [hunk 3]
```diff
@@ -308,6 +310,9 @@ Deepstream.prototype._init = function() {
 	this._recordHandler = new RecordHandler( this._options );
 	this._messageDistributor.registerForTopic( C.TOPIC.RECORD, this._recordHandler.handle.bind( this._recordHandler ) );
 
+	this._presenceHandler = new PresenceHandler( this._options, this._connectionEndpoint );
```
Comment (yasserf) [inline src/deepstream.io.js]: Lets be cheeky and put this on the `options` object instead to so we can call added/remove methods.

File: src/cluster/cluster-registry.js [hunk 4]
```diff
@@ -27,6 +27,7 @@ module.exports = class ClusterRegistry extends EventEmitter {
 	 */
 	constructor( options, connectionEndpoint ) {
 		super();
+		this.setMaxListeners( 20 );
```
Comment (AlexBHarley) [inline src/cluster/cluster-registry.js]: Was getting a `(node) warning: possible EventEmitter memory leak detected. 11 remove listeners added. Use emitter.setMaxListeners() to increase limit.`

File: src/message/connection-endpoint.js (hunk 1 above)
Comment (AlexBHarley) [inline src/message/connection-endpoint.js]: Can you elaborate on this? Not 100% sure what you mean

File: src/cluster/cluster-registry.js (hunk 4 above)
Comment (yasserf) [inline src/cluster/cluster-registry.js]: Yeah, that isn't the best approach since it will most likely also reach 20.

Can you out this in temporary:

``
process.on('warning', e => console.warn(e.stack));
```

and copy the stack trace?

Comment (AlexBHarley) [inline src/cluster/cluster-registry.js]: `(node) warning: possible EventEmitter memory leak detected. 11 remove listeners added. Use emitter.setMaxListeners() to increase limit.`
`Trace`
    `at ClusterRegistry.addListener (events.js:239:17)`
    `at new DistributedStateRegistry (C:\Users\Alex\dev\deepstream.io\src\cluster\distributed-state-registry.js:43:33)`
    `at new SubscriptionRegistry (C:\Users\Alex\dev\deepstream.io\src\utils\subscription-registry.js:31:32)`
    `at new PresenceHandler (C:\Users\Alex\dev\deepstream.io\src\presence\presence-handler.js:15:28)`
    `at Deepstream._init (C:\Users\Alex\dev\deepstream.io\src\deepstream.io.js:314:26)`
    `at Deepstream._checkReady (C:\Users\Alex\dev\deepstream.io\src\deepstream.io.js:345:8)`
    `at Deepstream._start (C:\Users\Alex\dev\deepstream.io\src\deepstream.io.js:174:7)`
    `at g (events.js:260:16)`
    `at emitNone (events.js:67:13)`
    `at emit (events.js:166:7)`

Comment (yasserf) [inline src/cluster/cluster-registry.js]: Interesting, it appears its because we have 11 distributed state registries. You can set the maxListeners to 11 then and we can add increase that each time to ensure nothing wierd is happening anywhere else

**Comment by yasserf** [comment]
Hey Alex,

In regards to cluster tests, we now are using exactly four clients for testing in order to cover all possible edge cases in different scenarios. Be good if I can show you edge cases we want to cover using that number over a call!

**Comment by jensvdh** [comment]
Hi. This commit has been merged into master. Does that mean a presence api is now available? If so where can I find the documentation for it? 

Thanks!!

**Comment by yasserf** [comment]
Hey!

We only merged in the server aspect, the client and documentation are still a WIP. Should be done by end of month!

**Comment by yasserf** [comment]
this is how the api will look like btw:

https://github.com/deepstreamIO/deepstream.io-client-js/pull/233

we just have a few conflicts in test scenarios that need to be sorted,
